onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,157 @@
+ * Get the max feature level supported by all target devices.
+ *
+ * @param nnapi nnapi-lib handle.
+ * @param device_handles target devices users want to use.
+ *
+ * @return The max feature level support by a set of devices.
+ *
+ * @exception-safe no throw.
+ */
+static int32_t GetDeviceFeatureLevelInternal(const ::NnApi* nnapi, const std::vector<ANeuralNetworksDevice*>& device_handles) {
+  if (!nnapi)
+    return 0;
+
+  int32_t target_feature_level = GetNNAPIRuntimeFeatureLevel(nnapi);
+
+  int64_t devices_feature_level = -1;
+
+  for (const auto* device_handle : device_handles) {
+    int64_t curr_device_feature_level = 0;
+    if (nnapi->ANeuralNetworksDevice_getFeatureLevel(device_handle, &curr_device_feature_level) != ANEURALNETWORKS_NO_ERROR) {
+      continue;
+    }
+
+    // we want to op run on the device with the highest feature level so we can support more ops.
+    // and we don't care which device runs them.
+    devices_feature_level = std::max(curr_device_feature_level, devices_feature_level);
+  }
+
+  // nnapi-cpu has the feature 1000
+  if ((devices_feature_level > 0) && (devices_feature_level < target_feature_level)) {
+    LOGS_DEFAULT(INFO) << "Changing NNAPI Feature Level " << target_feature_level
+                       << " to supported by target devices: " << devices_feature_level;
+
+    target_feature_level = static_cast<int32_t>(devices_feature_level);
+  }
+  return target_feature_level;
+}
+
+// get all target devices which satisfy the target_device_option
+// we will always put CPU device at the end if cpu is enabled
+Status GetTargetDevices(const ::NnApi* nnapi, TargetDeviceOption target_device_option,
+                        std::vector<ANeuralNetworksDevice*>& nnapi_target_devices, std::string& nnapi_target_devices_detail) {
+  // GetTargetDevices is only supported when NNAPI runtime feature level >= ANEURALNETWORKS_FEATURE_LEVEL_3
+  if (GetNNAPIRuntimeFeatureLevel(nnapi) < ANEURALNETWORKS_FEATURE_LEVEL_3)
+    return Status::OK();
+
+  const std::string nnapi_cpu("nnapi-reference");
+  uint32_t num_devices = 0;
+  RETURN_STATUS_ON_ERROR_WITH_NOTE(
+      nnapi->ANeuralNetworks_getDeviceCount(&num_devices), "Getting count of available devices");
+
+  int32_t cpu_index = -1;
+  for (uint32_t i = 0; i < num_devices; i++) {
+    ANeuralNetworksDevice* device = nullptr;
+    const char* device_name = nullptr;
+    int32_t device_type = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(
+        nnapi->ANeuralNetworks_getDevice(i, &device), "Getting " + std::to_string(i) + "th device");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi->ANeuralNetworksDevice_getName(device, &device_name),
+                                     "Getting " + std::to_string(i) + "th device's name");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi->ANeuralNetworksDevice_getType(device, &device_type),
+                                     "Getting " + std::to_string(i) + "th device's type");
+    bool device_is_cpu = nnapi_cpu == device_name;
```

Comment:
Would a comparison of the device_type with ANEURALNETWORKS_DEVICE_TYPE_CPU be better?

Reply:
Sounds good.  it's more effient.